components/higher-order/with-safe-timeout/index.js

```diff
@@ -0,0 +1,46 @@
+/**
+ * WordPress dependencies
+ */
+import { Component } from '@wordpress/element';
+
+/**
+ * Browser dependencies
+ */
+const { clearTimeout, setTimeout } = window;
+
+/**
+ * A higher-order component used to provide and manage delayed function calls
+ * that ought to be bound to a component's lifecycle.
+ *
+ * @param {Component} OriginalComponent   Component requiring setSafeTimeout
+ *
+ * @returns {Component}                   Wrapped component.
+ */
+function withSafeTimeout( OriginalComponent ) {
+	return class WrappedComponent extends Component {
+		constructor() {
+			super( ...arguments );
+			this.timeouts = [];
+			this.setSafeTimeout = this.setSafeTimeout.bind( this );
+		}
+
+		componentWillUnmount() {
+			this.timeouts.forEach( clearTimeout );
+		}
+
+		setSafeTimeout( fn, delay ) {
+			this.timeouts.push( setTimeout( fn, delay ) );
```

Comment:
Do you think we should remove finished timeouts? Or should we just let them pile up?

Reply:
I was picturing more short-lived instances, but good point: we should be defensive anyway.